google/cloud/spanner_dbapi/cursor.py

```diff
@@ -88,6 +89,20 @@ def execute(self, sql, args=None):
 
         self._res = None
 
+        if not self._connection.autocommit:
+            if (
+                not self.transaction
+                or self.transaction.committed
+                or self.transaction.rolled_back
+            ):
+                self.transaction = self._connection.session_checkout().transaction()
```

Comment:
The session that is checked out here, will never be returned to the pool. It is probably better to keep a reference to the current session and transaction in the `Connection` class, and to return the session to the pool when the active transaction on the connection is committed or rolled back.